### PR TITLE
Add retry to index read, shard read and schema read after write operation

### PR DESCRIPTION
**Pull request by reyreaud-l**
### What's being changed:

* Remove sleep in acceptance tests after write to ensure consistency @reyreaud-l 
* Add backoff on Index read, add a getter to avoid repeating the retry in multiple places @moogacs 
* Add backoff on shard read, add a getter to avoid repeating the retry in multiple places @moogacs 
* Add retry on read after write operation @reyreaud-l 
* Update some acceptance tests with minimal sleeps to ensure eventual consistency work in particular cases (killing nodes, etc...) @reyreaud-l 
* Fine tune the CI and ensure it passes @moogacs @reyreaud-l 

#### backoff usage pattern 

```go
backoff.Retry(func() error {
	// logic to retry and eventually return error if still exists
}, {retry count and sleep time usually every 50ms for 3 times })
```

### Review checklist

- [ ] Documentation has been updated, if necessary. Link to changed documentation:
- [ ] Chaos pipeline run or not necessary. Link to pipeline:
- [ ] All new code is covered by tests where it is reasonable.
- [ ] Performance tests have been run or not necessary.

### CODE REVIEW

**Comment by redouan-rhazouani** [comment]
Overall, the PR significantly improves eventual consistency, which is a great accomplishment. However, there are a couple of areas that still need refinement

**Review by redouan-rhazouani — CHANGES_REQUESTED** [review]

File: cloud/store/schema.go [hunk 1]
```diff
@@ -60,26 +62,28 @@ func (s *schema) Len() int {
 }
 
 func (s *schema) addClass(cls *models.Class, ss *sharding.State) error {
-	s.Lock()
```
[the remark this answers is not in the document]
Comment (moogacs) [reply cloud/store/schema.go:63]: write ops not retryable, it's the read only that is retryable , the `ReadMetaClass()` jsut does what was already before but with retries

File: cloud/store/schema.go [hunk 2]
```diff
@@ -355,30 +364,32 @@ func (s *schema) ShardReplicas(class, shard string) ([]string, error) {
 
 // TenantShard returns shard name for the provided tenant and its activity status
 func (s *schema) TenantShard(class, tenant string) (string, string) {
-	s.RLock()
-	defer s.RUnlock()
-
-	i := s.Classes[class]
-	if i == nil || !i.Sharding.PartitioningEnabled {
-		return "", ""
+	var exTenant, status string
```
[the remark this answers is not in the document]
Comment (moogacs) [reply cloud/store/schema.go:367]: the retry mechanism would be called at least once, if the data already there it won't affect the performance because it won't be retried any way

File: adapters/repos/db/repo.go [hunk 3]
```diff
@@ -196,65 +199,63 @@ type Config struct {
...
 	if err := index.drop(); err != nil {
 		db.logger.WithField("action", "delete_index").WithField("class", className).Error(err)
 	}
-	delete(db.indices, id)
 
-	db.promMetrics.DeleteClass(className.String())
-	return nil
+	delete(db.indices, indexID(className))
+
+	return db.promMetrics.DeleteClass(className.String())
```
[the remark this answers is not in the document]
Comment (moogacs) [reply adapters/repos/db/repo.go:258]: sure, `db.promMetrics.DeleteClass` return nil anyway but will change it to log instead 👍🏼

File: usecases/sharding/remote_index_incoming.go [hunk 4]
```diff
@@ -92,8 +92,8 @@ func NewRemoteIndexIncoming(repo RemoteIncomingRepo) *RemoteIndexIncoming {
 func (rii *RemoteIndexIncoming) PutObject(ctx context.Context, indexName,
 	shardName string, obj *storobj.Object,
 ) error {
-	index := rii.repo.GetIndexForIncoming(schema.ClassName(indexName))
-	if index == nil {
+	index, err := rii.repo.GetIndexForIncoming(schema.ClassName(indexName))
+	if err != nil {
 		return errors.Errorf("local index %q not found", indexName)
```
[the remark this answers is not in the document]
Comment (reyreaud-l) [reply usecases/sharding/remote_index_incoming.go:97]: I agree I think having the error is a beneficial reminder to check for it

File: test/acceptance/replication/crud_test.go [hunk 5]
```diff
@@ -332,6 +332,9 @@ func eventualReplicaCRUD(t *testing.T) {
 		helper.UpdateClass(t, pc)
 	})
 
+	// Ensure consistency on follower nodes by sleeping BEFORE shutting down the leader node
+	time.Sleep(3 * time.Second)
```
[the remark this answers is not in the document]
Comment (reyreaud-l) [reply test/acceptance/replication/crud_test.go:336]: I would to double check RAFT behaviour here. But in our tests this was the only way to solve this. Basically no data would be present on node_2 without a small sleep here. I'll re-check the logs and understand what is going on.

[the remark this answers is not in the document]
Comment (moogacs) [reply test/acceptance/replication/crud_test.go:336]: this one i think it's needed because the test actually just configured the number of replica from 1 to 2 so if we stopped 1 immediatly it won't have a chance to broadcast to 2 and it will end with error 
`broadcast: cannot reach enough replicas`

File: cloud/store/schema.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply cloud/store/schema.go:63]: Let's put it another way: `addClass()` is calling `ReadMetaClass()`, which retries the read even though it is not necessary, as explained above. To solve this, changes to write operations must be reverted in this case

File: cloud/store/schema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply cloud/store/schema.go:367]: The code logic appears to be: `determineObject -> Retry::TenantShard -> ReadMetaClass -> Retry::ReadClass`. As you can see, the second retry is embedded inside the first one. It would be better to retry the entire operation just once

File: test/acceptance/replication/crud_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (reyreaud-l) [reply test/acceptance/replication/crud_test.go:336]: But it should be applied in the log and the follower should apply it. Not sure how the process for that works so i need to look into it

File: cloud/store/schema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (moogacs) [reply cloud/store/schema.go:367]: i've updated the code and removed retrying `TenantShard`, it's resolved with a different comment

File: test/acceptance/replication/crud_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (moogacs) [reply test/acceptance/replication/crud_test.go:336]: that is my thoughts: the update will happen on Node one with configuration 
```go
pc.ReplicationConfig = &models.ReplicationConfig{
			Factor: 2,
		}
```

`node 1` given that the replication factor was 1 before, so the node won't update anything but in the middle of replication to `node 2` we shutdown the node and then ask `node 2(follower)` about the data while `node 1` is still down, i think the data will not be there yet until at least node1 is up again with note that we don't restart here until we ask node 2 we just shutdown node 1 and ask node 2 

in other words what we currently do 
update class replication factor  ->  shutdown node1 immediately after update success  ->  ask node 2 about the data -> start node 1 


because node 1 is down while we are asking node 2 about data not replicated yet

File: usecases/sharding/remote_index_incoming.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply usecases/sharding/remote_index_incoming.go:97]: I think returning an error for a non-existent index may not provide significant benefits, especially if we need to check for a non-nil pointer regardless. I am not opposed to this change but I would avoid it because it's not directly related to Raft, and altering it could lead to conflicts with the master branch

File: test/acceptance/replication/crud_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (reyreaud-l) [reply test/acceptance/replication/crud_test.go:336]: I agree here the consistency is not about the schema but about the actual data underneath as we discussed in a call earlier. It makes sense to keep the sleep IMO.

[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply test/acceptance/replication/crud_test.go:336]: The question arises: why was it functioning previously without a time delay? There seems to be something missing here. If you examine the sub-test "Configure classes to replicate to node 2," you'll notice the necessity to replicate the object on both nodes after a successful class update. The sleep afterward is not necessary because the main work occurs beforehand

File: cloud/store/schema.go [hunk 6]
```diff
@@ -226,13 +235,11 @@ type ClassInfo struct {
 }
 
 func (s *schema) ClassInfo(class string) (ci ClassInfo) {
-	s.RLock()
-	defer s.RUnlock()
-
-	i := s.Classes[class]
-	if i == nil {
+	i, err := s.ReadMetaClass(class)
```
Comment (redouan-rhazouani) [inline cloud/store/schema.go:238]: ClassInfo is used inside store.Execute() and hence should not retry ReadClass

File: usecases/sharding/remote_index_incoming.go [hunk 7]
```diff
@@ -92,8 +92,8 @@ func NewRemoteIndexIncoming(repo RemoteIncomingRepo) *RemoteIndexIncoming {
 func (rii *RemoteIndexIncoming) PutObject(ctx context.Context, indexName,
 	shardName string, obj *storobj.Object,
 ) error {
-	index := rii.repo.GetIndexForIncoming(schema.ClassName(indexName))
-	if index == nil {
+	index, err := rii.repo.GetIndexForIncoming(schema.ClassName(indexName))
```
Comment (redouan-rhazouani) [inline usecases/sharding/remote_index_incoming.go:95]: I think returning an error for a non-existent index may not provide significant benefits, especially if we need to check for a non-nil pointer regardless. I am not opposed to this change but I would avoid it because it's not directly related to Raft, and altering it could lead to conflicts with the master branch

**Review by redouan-rhazouani — CHANGES_REQUESTED** [review]

File: test/acceptance/replication/crud_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (reyreaud-l) [reply test/acceptance/replication/crud_test.go:336]: It was functionning before with a time delay on the UpdateClass just above https://github.com/weaviate/weaviate/pull/4095/files#diff-8f7596ee44f2376227bba6a51967324fde97a130afe64b71e66191a0cb046747L62.

Small but sufficient to make it work.

[the remark this answers is not in the document]
Comment (moogacs) [reply test/acceptance/replication/crud_test.go:336]: i think it was succeeding before because we had sleeps within `GetClass` and `UpdateClass`, so it's success was hidden under that sleeps of the Class CRUD

File: usecases/sharding/remote_index_incoming.go (hunk 7 above)
Comment (moogacs) [reply usecases/sharding/remote_index_incoming.go:95]: we could revisit it  again if we are not in favor in a different PR

File: cloud/store/schema.go [hunk 8]
```diff
@@ -103,6 +107,9 @@ func (s *schema) addProperty(class string, p models.Property) error {
 		return errClassNotFound
 	}
 
+	s.Lock()
```
[the remark this answers is not in the document]
Comment (moogacs) [reply cloud/store/schema.go:110]: good catch

File: test/acceptance/replication/crud_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (redouan-rhazouani) [reply test/acceptance/replication/crud_test.go:336]: The retry should handle the delay, but it seems something may be missing. We can investigate further and create a task to address if necessary.

File: cloud/store/schema.go [hunk 9]
```diff
@@ -66,20 +68,22 @@ func (s *schema) addClass(cls *models.Class, ss *sharding.State) error {
 	if exists {
 		return errClassExists
 	}
+
 	s.Classes[cls.Class] = &metaClass{*cls, *ss}
 	return nil
 }
 
-func (s *schema) updateClass(u *models.Class, ss *sharding.State) error {
+func (s *schema) updateClass(cls *models.Class, ss *sharding.State) error {
```
[the remark this answers is not in the document]
Comment (moogacs) [reply cloud/store/schema.go:76]: it should be already updated

File: test/acceptance/replication/crud_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (reyreaud-l) [reply test/acceptance/replication/crud_test.go:336]: Created a follow-up task: https://semi-technology.atlassian.net/browse/WEAVIATE-701

**Comment by reyreaud-l** [comment]
The last failure in CI has been tackled on main and is unrelated to that change. The job should be successful once we rebase poc/schema-v2 on main.

File: adapters/handlers/rest/clusterapi/indices.go [hunk 10]
```diff
@@ -358,8 +358,8 @@ func (i *indices) postObjectBatch(w http.ResponseWriter, r *http.Request,
 		return
 	}
 
-	index := i.db.GetIndexForIncoming(entschema.ClassName(class))
-	if index == nil {
+	index, err := i.db.GetIndexForIncoming(entschema.ClassName(class))
```
Comment (redouan-rhazouani) [inline adapters/handlers/rest/clusterapi/indices.go:361]: Considering the point I mentioned previously, I recommend reverting this change. 
First, it appears unnecessary and are not related directly to RAFT. In addition, it could lead to conflicts with the main branch. Let's prioritise changes that are essential to minimise friction with the main branch. Feel free to propose this change again, once merge RAFT branch into main branch

File: adapters/repos/db/index.go [hunk 11]
```diff
@@ -1901,13 +1905,19 @@ func (i *Index) findUUIDs(ctx context.Context,
...
-		if shard := i.localShard(shardName); shard != nil {
-			res, err = shard.FindUUIDs(ctx, filters)
-		} else {
-			res, err = i.remote.FindUUIDs(ctx, shardName, filters)
-		}
+		err = backoff.Retry(func() error {
+			if shard := i.localShard(shardName); shard != nil {
+				res, err = shard.FindUUIDs(ctx, filters)
+			} else {
+				res, err = i.remote.FindUUIDs(ctx, shardName, filters)
```
Comment (redouan-rhazouani) [inline adapters/repos/db/index.go:1912]: `remote.FindUUIDs()` is making a call to a remote node. I don't think we need to retry here. Our focus is on managing eventual consistency on the host node

File: adapters/repos/db/replication.go [hunk 12]
```diff
@@ -367,17 +369,14 @@ func (db *DB) DigestObjects(ctx context.Context,
 	class, shardName string, ids []strfmt.UUID,
 ) (result []replica.RepairResponse, err error) {
 	index := db.GetIndex(schema.ClassName(class))
-	if index == nil {
```
Comment (redouan-rhazouani) [inline adapters/repos/db/replication.go:370]: Is removing the check for nil index intended?

**Review by redouan-rhazouani — CHANGES_REQUESTED** [review]
PR looking much better now. However, there are still some important issues that need to be addressed.

File: adapters/repos/db/replication.go (hunk 12 above)
Comment (moogacs) [reply adapters/repos/db/replication.go:370]: nope wasn't intended, got catch

File: adapters/repos/db/index.go (hunk 11 above)
Comment (moogacs) [reply adapters/repos/db/index.go:1912]: we need to retry the localShard though

Comment (reyreaud-l) [reply adapters/repos/db/index.go:1912]: We can isolate the localShard call in it's own retry and then as a follow up either call local `FindUUIDs` or call it on remote.

Comment (moogacs) [reply adapters/repos/db/index.go:1912]: yeah on it

**Comment by reyreaud-l** [comment]
Closed in favor of https://github.com/weaviate/weaviate/pull/4148
